tsdb/ooo_head_read.go

```diff
@@ -174,6 +186,60 @@ type multiMeta struct {
 	metas          []chunks.Meta
 }
 
+// Wrapper to return a chunk where the iterator clears
+// counter resets.
+type hintClearedChunk struct {
+	chunkenc.Chunk
+}
+
+func (h hintClearedChunk) Iterator(it chunkenc.Iterator) chunkenc.Iterator {
+	if itt, ok := it.(hintClearedIterator); ok {
+		itt.it = h.Chunk.Iterator(itt.it)
+		return itt
+	}
+	return &hintClearedIterator{it: h.Chunk.Iterator(nil)}
+}
```

Comment:
I think we also need to implement `Bytes()` directly in `hintClearedChunk` for the cases where the bytes of the chunk are directly read rather than samples read via the iterator. For example, in compaction: https://github.com/prometheus/prometheus/blob/cde42f30e9c91a23458661ee33d47f828fa63b43/tsdb/chunks/chunks.go#L526
Otherwise the header won't be cleared correctly in all cases.

To be very safe I'd prefer to not embed `chunkenc.Chunk` and instead have it as a named field in `hintClearedChunk` and add panics for the methods that we're sure aren't called.

Reply:
Resolved as headers are now set in ToEncodedChunks()